text/0000-buildpack-registry-search-api.md

```diff
@@ -0,0 +1,172 @@
+  {
+      "description": "The Command Function Buildpack is a Cloud Native Buildpack V3 that provides riff Command Function Invoker to functions",
+      "license": "MIT",
+      "ns":"projectriff",
+      "name":"command-function",
+      "latest_version": "1.4.1",
+      "versions": "/buildpacks/1/versions",
+      "weekly_downloads": 8993,
+      "issues": 44,
+      "addr":"gcr.io/projectriff/command-function@sha256:99f9054abb73635a9b251b61d3627a8ff86508c767f9d691c426d45e8758596f"
+  }
+  ```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+In the initial implementation, we can take advantage of the existing [registry](https://github.com/buildpacks/registry-index) repository to pull buildpack data from.  Pack already performs a clone/pull on this repo for its registry cache.  To keep ensure up-to-date buildpack data, we'd have a polling loop that performs a `git pull`.  This polling would take place in a separate background process.  
+
+Each time the local repository has been updated via `git pull`, buildapack data will be processed/normalized into a single JSON object, where plain text searches can be used against it.   Fields in this JSON object will be re-indexed as searchable fields, which will be compared against the search text during the retrieval algorithm.
+
+In addition, a server process will expose endpoints and handle incoming request for the GET endpoints mentioned earlier.  For search requests, the text will be extracted from the `query` query parameter and used to search against the normalized buildpacks index.   Results will be added to the server response, as a list of JSON objects.
+
+*Note:  The Distribution Team maintainers can manage this service, and could even set-up a pager (but only best effort, and during working hours).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+- more vulnerable due to Github outages
+- less optimal than using actual DB technology
+- constraint by what's in the index registry repository.  We might want to augment the metadata, with new fields that don't makes sense to have in the index repository.  I think this maybe an issue for the `/buildpacks/1` endpoint that *could* have download data, etc. things that don't belong in the actual repo.
+- without a DB we can't easily do things like ORDER_BY, or use ASC|DESC
+- someone has to maintain the service
+
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+
+  - Do nothing
+  
+    We already have buildpack metadata stored in the [registry](https://github.com/buildpacks/registry-index).  
+    Customers can simply explore this repository for buildpack information.  Perhaps someone else in the community will
+    come up with their own solution that can be donated to CNB.
+
+  - Having a separate, independent service responsible for:
+    1. pulling from the registry-index repository
+    2. augmenting buildpack data
+    3. persisting it to a hosted PostgresDB  
+
```

Comment:
Another option: we could implement search purely on the client side (with `pack` for example), instead of providing a service. This would have the benefit of searching across all user-configured registries not just https://github.com/buildpacks/registry-index.

Reply:
We could implement this entirely on the client side for both `pack` and a website.  But I don't like the tight coupling it would create for the web-based client.   However, I think we should definitely do this for `pack`.   Along these lines, it sounds like from the last WG, we might also want to consider a registry spec, so that `pack` would work against other private registries.  For the web case, I see it being very specific for CNB.   However, with loser coupling, customers could more easily re-use our API service and re-skin it with their own search dashboard (website).